### ISSUE TITLE
[question] passing pointers to/from wasm function

### ISSUE THREAD

**Issue by konsumer** [wasm3/wasm3#419]
Forgive me if this is a  dumb question, but I can't seem to figure out the answer from any of the examples or looking at the source, but C is not my strong-suit.

I am trying to pass strings to/from WASM made in assemblyscript. On the WASM-side, I am converting everything to null-terminated UTF8 strings, and I can call functions in the host, passing strings.

In AS, I have a simple function that looks like this:

```ts
export function stringinout_utf8(n: ArrayBuffer) : ArrayBuffer {
  const name = String.UTF8.decode(n, true)
  return String.UTF8.encode(`Hello ${name}`, true)
}
```

This works in a nodejs host, like this:

```js
// ArrayBuffer -> String
export function ab2str (buf) {
  return String.fromCharCode.apply(null, new Uint8Array(buf).slice(0, -1))
}

// String -> ArrayBuffer
export function str2ab (str) {
  const buf = new ArrayBuffer(str.length + 1)
  buf[str.length] = 0
  const bufView = new Uint8Array(buf)
  for (let i = 0, strLen = str.length; i < strLen; i++) {
    bufView[i] = str.charCodeAt(i)
  }
  return buf
}

const { memory, __new, stringinout_utf8 } = wasmModule.instance.exports

function __liftBuffer (pointer) {
  if (!pointer) return null
  return memory.buffer.slice(pointer, pointer + new Uint32Array(memory.buffer)[pointer - 4 >>> 2])
}

function __lowerBuffer (value) {
  if (value == null) return 0
  const pointer = __new(value.byteLength, 1) >>> 0
  new Uint8Array(memory.buffer).set(new Uint8Array(value), pointer)
  return pointer
}

const pointerInName = __lowerBuffer(str2ab('World'))
const pointerOutRet = stringinout_utf8(pointerInName)
const greet = ab2str(__liftBuffer(pointerOutRet))
assert.strictEqual(greet, 'Hello World')
```

Essentially, I inject the value into the wasm memory, then give it a pointer for the argument.  It returns a pointer to it's own memory for response, then I pull that out.

My goal is to be able to write code in many different languages that compile to wasm, but run them in the same C-based runtime. Should I pass strings back/forth some other way?

I tried sort of "porting" the nodejs host code to C, but I don't think it's quite right:

```c
uint32_t wallocate(uint32_t size, uint32_t id) {
  m3_CallV(__new, size, id);
  uint32_t ptr = 0;
  m3_GetResultsV(__new, &ptr);
  return ptr;
}

static M3Function* __new;
static M3Function* stringinout_utf8;
m3_FindFunction(&__new, runtime, "__new");
m3_FindFunction(&stringinout_utf8, runtime, "stringinout_utf8");

char* name = "World";
uint32_t ptrName = wallocate(sizeof(name), 1);
m3_CallV(stringinout_utf8, ptrName);
char* greet;
m3_GetResultsV(stringinout_utf8, greet);
printf("GREET: %s\n", greet);
```

Basically, `wallocate` is a lil wrapper around `__new(size, id)` in the WASM, where id of `1` means `ArrayBuffer`.

I get
```
GREET: 0
```

**Comment by konsumer** [comment]
This might seem like the same question as #340 but now I want to return a string to WASM, instead of receive one. The encoding part seems fine, but this simple example throws `out of bounds memory access`:

```c
static m3ApiRawFunction (ret_string) {
  m3ApiReturnType (char*);
  char* out = "you want some string";
  m3ApiReturn(out);
  m3ApiSuccess();
}
```

**Comment by konsumer** [comment]
I may have made progress, trying to sort of copy what assemblyscript-generated wrapper does, but I am stuck again.

I made this WASM-side:

```ts
export function wmalloc(size: usize): usize {
  return __pin(__new(size, 1))
}

// free an ArrayBuffer
export function wfree(ptr: usize): void {
  __unpin(ptr)
}
```
and this C-side:

```c
// take a wasm-pointer to a buffer and return it
char* liftBuffer(uint32_t ptr) {
  return (char*) m3ApiOffsetToPtr(wPointer);
}

// take a buffer and put it in wasm-memory, return pointer
uint32_t lowerBuffer(char* buffer) {
  size_t s = sizeof(buffer);
  char* wPointer = (char*) wmalloc(s);
  char* wpBuffer = (char*) m3ApiOffsetToPtr(wPointer);
  memcpy(wpBuffer, buffer, s);
  return wPointer;
}
```

but I get this error:

```
/Users/konsumer/Desktop/otherdev/wasm-strings-example/chost/src/main.c:24:18: error: use of undeclared identifier '_mem'
  return (char*) m3ApiOffsetToPtr(wPointer);
```

How do I make sure `_mem` is defined?

**Comment by konsumer** [comment]
Ok, I think I answered all my own questions, after lots of thrashing around with code. Sorry for the noise.

For others:

You can get `_mem` like this, in places where it's not defined:

```c
M3Memory* _mem = &runtime->memory;
```

This works to return a string to wasm:

```c
static m3ApiRawFunction (test_string_get) {
  printf("test_string_get was called.\n");
  m3ApiReturnType (uint32_t);

  // example of what you want to return
  char* buffer = "Hello from C host (test_string_get)";

  // lowerBuffer
  uint32_t wPointer;
  size_t s = strlen(buffer) + 1;
  null0_check_wasm3(m3_CallV (wmalloc, s));
  m3_GetResultsV(wmalloc, &wPointer);
  char* wBuffer = m3ApiOffsetToPtr(wPointer);
  memcpy(wBuffer, buffer, s);

  m3ApiReturn(wPointer);
  m3ApiSuccess();
}
```

`wmalloc` looks like this (WASM-side), in assemblyscript:

```ts
// malloc for ArrayBuffer
export function wmalloc(size: usize): usize {
  return __pin(__new(size, 1))
}
```

This works to take strings as params from WASM:

```c
static m3ApiRawFunction (null0_log) {
  m3ApiGetArgMem(const char*, message);
  printf("Log from WASM: %s\n", message);
  m3ApiSuccess();
}
```

**Comment by konsumer** [comment]
Still not sure how to access wasm memory, outside of a `m3ApiRawFunction`.

I tried this:

```c
M3Memory* _mem = &runtime->memory;
```

**Comment by max22-** [comment]
I've spent some time looking for it but i think it is this function : 

`
uint8_t * m3_GetMemory (IM3Runtime i_runtime, uint32_t * o_memorySizeInBytes, uint32_t i_memoryIndex);
`

I tested reading data, it works. (not writing yet, but it should work?)

[link to the header file](https://github.com/wasm3/wasm3/blob/139076a98b8321b67f850a844f558b5e91b5ac83/source/wasm3.h#L215)

[example of usage](https://github.com/wasm3/wasm3/blob/139076a98b8321b67f850a844f558b5e91b5ac83/platforms/app/main.c#L443)
